module/input/PS3Controller/src/PS3Controller.cpp

```diff
@@ -0,0 +1,244 @@
+                            emit(std::make_unique<LeftJoystickButton>(timestamp, value > 0, is_init));
+                            break;
+                        case Button::JOYSTICK_RIGHT:
+                            emit(std::make_unique<RightJoystickButton>(timestamp, value > 0, is_init));
+                            break;
+                        case Button::DPAD_UP:
+                            emit(
+                                std::make_unique<DPadButton>(timestamp, DPadButton::Direction::UP, value > 0, is_init));
+                            break;
+                        case Button::DPAD_DOWN:
+                            emit(std::make_unique<DPadButton>(
+                                timestamp, DPadButton::Direction::DOWN, value > 0, is_init));
+                            break;
+                        case Button::DPAD_LEFT:
+                            emit(std::make_unique<DPadButton>(
+                                timestamp, DPadButton::Direction::LEFT, value > 0, is_init));
+                            break;
+                        case Button::DPAD_RIGHT:
+                            emit(std::make_unique<DPadButton>(
+                                timestamp, DPadButton::Direction::RIGHT, value > 0, is_init));
+                            break;
+                    }
+                }
+                else if (is_axis) {
+                    float normalised =
+                        static_cast<float>(value) / static_cast<float>(std::numeric_limits<int16_t>::max());
+                    switch (static_cast<Axis>(event_buffer.front())) {
+                        case Axis::JOYSTICK_LEFT_HORIZONTAL:
+                            emit(std::make_unique<LeftJoystick>(
+                                timestamp, LeftJoystick::Direction::HORIZONTAL, normalised, is_init));
+                            break;
+                        case Axis::JOYSTICK_LEFT_VERTICAL:
+                            emit(std::make_unique<LeftJoystick>(
+                                timestamp, LeftJoystick::Direction::VERTICAL, normalised, is_init));
+                            break;
+                        case Axis::L2: emit(std::make_unique<L2Trigger>(timestamp, normalised, is_init)); break;
+                        case Axis::JOYSTICK_RIGHT_HORIZONTAL:
+                            emit(std::make_unique<RightJoystick>(
+                                timestamp, RightJoystick::Direction::HORIZONTAL, normalised, is_init));
+                            break;
+                        case Axis::JOYSTICK_RIGHT_VERTICAL:
+                            emit(std::make_unique<RightJoystick>(
+                                timestamp, RightJoystick::Direction::VERTICAL, normalised, is_init));
+                            break;
+                        case Axis::R2: emit(std::make_unique<R2Trigger>(timestamp, normalised, is_init)); break;
+                        case Axis::ACCELEROMETER_X: /* TODO: Handle this. */ break;
```

Comment:
This comes from the other io right? So unless you’re moving this into its own function to process genetic joystick event structs (not a bad idea) you should just be able to delete these

Reply:
I think I put them there to shut the compiler up about not handle enum cases

Reply:
`default:` ?

Reply:
I retract my previous comment. It's probably a good idea in case other buttons are added. But I'd remove the TODO comment since we don't want to todo it here :P